src/server/frontend/wayland/wayland_connector.cpp

```diff
@@ -1967,62 +1954,207 @@ class WlShellSurface : public wayland::ShellSurface
     }
 
     void set_transient(
-        struct wl_resource* /*parent*/,
-        int32_t /*x*/,
-        int32_t /*y*/,
+        struct wl_resource* parent,
+        int32_t x,
+        int32_t y,
         uint32_t /*flags*/) override
     {
+        auto const session = session_for_client(client);
+        auto* tmp = wl_resource_get_user_data(parent);
+        auto& parent_surface = *static_cast<WlSurface*>(tmp);
+
+        if (surface_id.as_value())
+        {
+            shell::SurfaceSpecification new_spec;
+            new_spec.parent_id = parent_surface.surface_id;
+            new_spec.aux_rect = Rectangle{{X{x},    Y{y}},
+                                          {Width{}, Height{}}};
+            new_spec.surface_placement_gravity = mir_placement_gravity_northwest;
+            new_spec.aux_rect_placement_gravity = mir_placement_gravity_southeast;
+            new_spec.placement_hints = mir_placement_hints_slide_x;
+            new_spec.aux_rect_placement_offset_x = 0;
+            new_spec.aux_rect_placement_offset_y = 0;
+            shell->modify_surface(session, surface_id, new_spec);
+        }
+        else
+        {
+            params.parent_id = parent_surface.surface_id;
+            params.aux_rect = Rectangle{{x, y}, {}};
+            params.surface_placement_gravity = mir_placement_gravity_northwest;
+            params.aux_rect_placement_gravity = mir_placement_gravity_southeast;
+            params.placement_hints = mir_placement_hints_slide_x;
+            params.aux_rect_placement_offset_x = 0;
+            params.aux_rect_placement_offset_y = 0;
+        }
     }
 
     void set_fullscreen(
         uint32_t /*method*/,
         uint32_t /*framerate*/,
         std::experimental::optional<struct wl_resource*> const& output) override
     {
-        mir::shell::SurfaceSpecification mods;
-        mods.state = mir_window_state_fullscreen;
-        if (output)
+        if (surface_id.as_value())
```

Comment:
surface_id can not have a value? /me raises eyebrow.

Reply:
surface_id has a value of 0 until we create a window in the Mir scene, after which it identifies the window. (I'll listen to ideas on how to manage this state better.)